homeassistant/components/shelly/__init__.py

```diff
@@ -302,11 +287,19 @@ async def async_unload_entry(hass: HomeAssistant, entry: ShellyConfigEntry) -> b
         hass, DOMAIN, PUSH_UPDATE_ISSUE_ID.format(unique=entry.unique_id)
     )
 
-    if entry.data.get(CONF_SLEEP_PERIOD):
-        platforms = BLOCK_SLEEPING_PLATFORMS
+    platforms = PLATFORMS
+    coordinator: ShellyBlockCoordinator | ShellyRpcCoordinator | None
+
+    if get_device_entry_gen(entry) in RPC_GENERATIONS:
+        coordinator = entry.runtime_data.rpc
+        if entry.data.get(CONF_SLEEP_PERIOD):
```

Comment:
Another option is to store platforms on runtime_data so its only worked out at setup time, and unload can look at `runtime_data.platforms`